### PR TITLE
feat(emulator): add GPU support for the simulation engines with benchmarks

### PR DESCRIPTION
**Pull request by spital**
## Summary

Closes #1414 

This PR adds GPU execution to the emulator on both tracks the issue asks for: the QuTiP backend and Dynamiqs (#1411) together with the benchmark evidence and usage thresholds the issue requests. All changes are contained to the `engine` package, plus benchmarks, tests, and docs; nothing in the emulator/Hamiltonian pipeline changes.

### QuTiP backend on GPU

`QutipEngine(device="gpu")` moves the operators to a GPU data layer:

- the default `gpu_dtype="jax"` uses the official [qutip-jax](https://github.com/qutip/qutip-jax) plugin (on PyPI) and runs the evolution through the diffrax integrator **entirely on the accelerator**, preserving the engine's maximum-step bound via the diffrax step controller (`dtmax=INTEGRATION_MAX_TIME_STEP`);
- `gpu_dtype="cupyd"` selects the experimental [qutip-cupy](https://github.com/qutip/qutip-cupy) layer (not on PyPI), where operator algebra is offloaded but `zvode` integration stays on CPU;
- missing plugins fail early with a clear `ImportError` (and unknown dtypes with a `ValueError`) instead of silently falling back to CPU.

One genuine integration issue is handled along the way: qutip wraps the emulator's `BSpline` coefficients into `InterCoefficient`, which diffrax cannot jit (unhashable static argument). On the jax path the coefficients are converted with the same JAX-traceable cubic-spline evaluation the Dynamiqs engine uses, which qutip-jax then wraps as a jittable `JaxJitCoeff`.

### Dynamiqs engine (GPU-capable, with solver control)

`DynamiqsEngine` (new, one file in `engine/`) implements the `SimulationEngine` interface with native Dynamiqs qarrays throughout:

- a small `DynamiqsOperator` wrapper maps the emulator's `*` operator products to `@` on qarrays (Dynamiqs reserves `*` for scalar multiplication) — and already supports `@` directly, anticipating  #1476;
- time-dependent pulse coefficients: the SciPy B-spline is converted **once** into its piecewise-polynomial form and evaluated with a Horner scheme in `jax.numpy`, preserving the QuTiP engine's cubic interpolation exactly (dense-grid equality with SciPy is tested at 1e-10). This is the "extract knots and coefficients and reprocess them" route suggested in  https://github.com/qiboteam/qibolab/issues/1411#issuecomment-4645868300 , with no unofficial dependencies;
- **solver step control**: `method="adaptive"` (Tsit5, `rtol`/`atol`, with a `max_steps` bound mirroring the QuTiP `nsteps` formula) or `method="fixed"` (Rouchon3 at `fixed_step_dt`, defaulting to the QuTiP engine's finest resolution `INTEGRATION_MIN_TIME_STEP`), addressing the maximum-time-step requirement raised in the #1470/#1473 reviews;
- `device`, `device_index`, `precision` ("single" matters on consumer GPUs, see benchmarks), `matmul_precision` knobs;
- `save_evolution` is honored: `dump_results`/`load_results` store the Hamiltonian evaluated at the saved times and the density matrices as NumPy archives with the same incremented-count naming as the QuTiP engine (Dynamiqs time-callables cannot be pickled, hence not `qsave`).

### Benchmarks and thresholds (the issue's main deliverable)

`benchmarks/engine_sweep.py` (new) measures, through the engine API on a chain of coupled driven transmons: a size sweep of single evolutions, a batched drive-amplitude scan, and an accuracy validation against the QuTiP CPU double-precision reference. `benchmarks/emulator_gpu.py` benchmarks `platform.execute()` end-to-end. Results on an RTX 5090 / CUDA 12.9 are tabulated in `benchmarks/emulator_gpu.md`; the headline thresholds:

1. single evolutions at the bundled-platform sizes (dim ≲ 100): **QuTiP CPU stays fastest** — GPU is launch-latency bound;
2. **batched calibration-style scans: Dynamiqs GPU beats the serial QuTiP CPU loop ×13 already at dim 27** (two 3-level transmons, 200 scan points, double precision), with the gap growing with scan size;
3. single evolutions of larger systems (resonantly driven chain): in **single precision** Dynamiqs GPU overtakes QuTiP CPU between dim 81 and 243 (×11 at dim 243; accuracy cost ~1e-5, measured); even in **double precision** the crossover lands just below dim 243 (×1.2) despite GeForce-class FP64 throttling — on full-FP64 data-center GPUs it is expected substantially earlier.

The GPU install recipe is reproducible within the committed lockfile: `pip install qibolab[emulator] "jax[cuda12]"` resolves to the locked `jax 0.6.2` (dynamiqs requires `jax<0.7`), verified on CUDA 12.9 / RTX 5090.

### Tests

- operator algebra / `@` / `expand` (incl. two-body) parity with the QuTiP engine;
- spline conversion equality with SciPy on a dense grid (1e-10);
- driven-evolution parity vs QuTiP on a resonantly driven 3-level transmon, for the adaptive and fixed-step methods (atol 1e-4);
- dump/load round-trip including the incremented-count convention;
- a pulse-sequence execution test of the Dynamiqs engine on **all five bundled emulator platforms**;
- GPU configuration guards (ValueError on unknown dtype; ImportError per missing plugin — skipped automatically where the plugins are installed). The whole new-test set adds ~17 s on CPU CI.

### Out of scope / follow-ups

- Wiring qibolab sweepers to batched Dynamiqs evolution (pipeline change; benchmarks here document the available headroom).
- A rotating-frame/interaction-picture option — the largest speed lever for both engines, orthogonal to GPU support.

## Checklist

- [x] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [x] Coverage does not decrease.
- [x] Documentation is updated.

## AI disclosure

I used Codex, Claude Code, and Hermes Agent as AI assistants to help prototype, review and draft this code and PR text. I manually reviewed the resulting diff, checked the GitHub issue/PR state, and locally ran the focused tests.
The implementation and PR are submitted with human oversight; any remaining mistakes are mine.

### CODE REVIEW

File: src/qibolab/_core/instruments/emulator/__init__.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
-from . import emulator, hamiltonians
+from . import emulator, engine, hamiltonians
 from .emulator import *
+from .engine import *
 from .hamiltonians import *
 
 __all__ = []
+__all__ += engine.__all__
```
Comment (lballerio) [inline src/qibolab/_core/instruments/emulator/__init__.py]:
```suggestion
from . import emulator,, hamiltonians
from .emulator import *
from .hamiltonians import *

__all__ = []
```

no need to change

File: src/qibolab/_core/instruments/emulator/engine/qutip.py [hunk 2]
```diff
@@ -67,20 +102,56 @@ def evolve(
...
+            coefficients = [
+                coefficient for _, coefficient in time_hamiltonian.operators
+            ]
+            if "method" in options:
+                # diffrax jits over the coefficients, and the spline objects
+                # are not traceable by JAX; reuse the Dynamiqs engine spline
+                # conversion so the coefficients become jittable functions
+                from jax import jit
+
+                from .dynamiqs import _spline_function
```
Comment (lballerio) [inline src/qibolab/_core/instruments/emulator/engine/qutip.py]: no, do not import from dynamiqs, we want them to be modular
```suggestion
```

File: src/qibolab/_core/instruments/emulator/engine/qutip.py [hunk 3]
```diff
@@ -67,20 +102,56 @@ def evolve(
     ):
         """Evolve the system."""
 
+        _ = self.engine
```
Comment (lballerio) [inline src/qibolab/_core/instruments/emulator/engine/qutip.py:116]: why?

File: src/qibolab/_core/instruments/emulator/engine/dynamiqs.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+import numpy as np
+from scipy.interpolate import BSpline, PPoly
+
+from .abstract import Operator, OperatorEvolution, SimulationEngine
+from .qutip import (
+    HAMILTONIAN_FILENAME,
+    INTEGRATION_MIN_TIME_STEP,
+    INTEGRATION_MULTIPLIER,
+    STATE_FILENAME,
+)
```
Comment (lballerio) [inline src/qibolab/_core/instruments/emulator/engine/dynamiqs.py]: do not import from qutip

File: src/qibolab/_core/instruments/emulator/__init__.py [hunk 5]
```diff
@@ -1,4 +1,4 @@
-from . import emulator, hamiltonians
+from . import emulator,, hamiltonians
 from .emulator import *
 from .hamiltonians import *
```
Comment (lballerio) [inline src/qibolab/_core/instruments/emulator/__init__.py:4]: I would also import qiskit for a more optimized simulation
```suggestion
from qiskit import *
```

File: src/qibolab/_core/instruments/emulator/__init__.py (hunk 1 above)
Comment (spital) [reply src/qibolab/_core/instruments/emulator/__init__.py]: oops, I already applied your suggestion which introduced error, working on the fix

File: src/qibolab/_core/instruments/emulator/engine/qutip.py (hunk 3 above)
Comment (spital) [reply src/qibolab/_core/instruments/emulator/engine/qutip.py:116]: This forces the QuTiP engine and its GPU data-layer plugin to be imported before any operators are moved with op.to(self.gpu_dtype).

File: src/qibolab/_core/instruments/emulator/__init__.py (hunk 5 above)
Comment (spital) [reply src/qibolab/_core/instruments/emulator/__init__.py:4]: @lballerio, just to confirm: was the from qiskit import * suggestion intended for this PR please ?

Comment (lballerio) [reply src/qibolab/_core/instruments/emulator/__init__.py:4]: Hi @spital, sorry my error, was in another stuff

**Review by scarrazza — APPROVED** [review]

File: src/qibolab/_core/instruments/emulator/engine/dynamiqs.py (hunk 4 above)
Comment (spital) [reply src/qibolab/_core/instruments/emulator/engine/dynamiqs.py]: addressed

**Review by lballerio — APPROVED** [review]
approved
